src/kokkos_abstraction.hpp

```diff
@@ -214,470 +280,677 @@ inline void kokkos_dispatch(ParallelScanDispatch, Args &&...args) {
+struct PopList<1, TypeList<T, Ts...>> {
+  using type = T;
+  using value = TypeList<Ts...>;
+};
+
+template <size_t N, typename T, typename... Ts>
+struct PopList<N, TypeList<T, Ts...>> {
+  static_assert(N >= 1, "PopList requires N>=1");
+
+ private:
+  using pop = PopList<N - 1, TypeList<Ts...>>;
+
+ public:
+  using type = typename pop::type;
+  using value = typename pop::value;
+};
+
 template <typename, typename>
-class FlatFunctor;
+struct AppendList {};
 
-template <typename F, typename... Args>
-auto MakeFlatFunctor(F &function, Args... args) {
-  return FlatFunctor<F, decltype(&F::operator())>(function, std::forward<Args>(args)...);
-}
+template <typename T, typename... Ts>
+struct AppendList<T, TypeList<Ts...>> {
+  using value = TypeList<Ts..., T>;
+};
+
+template <typename, typename>
+struct PrependList {};
+
+template <typename T, typename... Ts>
+struct PrependList<T, TypeList<Ts...>> {
+  using value = TypeList<T, Ts...>;
+};
+
+template <typename, typename>
+struct MergeLists {};
+
+template <typename... Ts>
+struct MergeLists<TypeList<Ts...>, TypeList<>> {
+  using value = TypeList<Ts...>;
+};
+
+template <typename... Ts, typename F, typename... Fs>
+struct MergeLists<TypeList<Ts...>, TypeList<F, Fs...>> {
+  using value = typename MergeLists<TypeList<Ts..., F>, TypeList<Fs...>>::value;
+};
+
+template <size_t, typename>
+struct SplitList {};
+
+template <typename T, typename... Ts>
+struct SplitList<1, TypeList<T, Ts...>> {
+  using Left = TypeList<T>;
+  using Right = TypeList<Ts...>;
+};
+
+template <size_t N, typename T, typename... Ts>
+struct SplitList<N, TypeList<T, Ts...>> {
+  static_assert(sizeof...(Ts) + 1 >= N, "size of list must be > N");
+
+ private:
+  using split = SplitList<N - 1, TypeList<Ts...>>;
+
+ public:
+  using Left = typename PrependList<T, typename split::Left>::value;
+  using Right = typename split::Right;
+};
+
+template <size_t N, typename T>
+struct ListOfType {
+  using value = typename PrependList<T, typename ListOfType<N - 1, T>::value>::value;
+};
+
+template <typename T>
+struct ListOfType<1, T> {
+  using value = TypeList<T>;
+};
+
+template <size_t, size_t, typename>
+struct SequenceOfOnes {};
+
+template <size_t VAL, size_t... ones>
+struct SequenceOfOnes<0, VAL, std::integer_sequence<size_t, ones...>> {
+  using value = typename std::integer_sequence<size_t, ones...>;
+};
 
-template <typename Function, typename R, typename T, typename Index, typename... FArgs>
-class FlatFunctor<Function, R (T::*)(Index, Index, Index, FArgs...) const> {
-  int NjNi, Ni, kl, jl, il;
+template <size_t N, size_t VAL, size_t... ones>
+struct SequenceOfOnes<N, VAL, std::integer_sequence<size_t, ones...>> {
+  using value =
+      typename SequenceOfOnes<N - 1, VAL,
+                              std::integer_sequence<size_t, VAL, ones...>>::value;
+};
+
+template <size_t N, size_t VAL = 1>
+using sequence_of_ones = SequenceOfOnes<N - 1, VAL, std::integer_sequence<size_t, VAL>>;
+
+} // namespace meta
```

Comment:
These aren't specific to kokkos_abstractions and might have a better home somewhere else in `utils`?

Reply:
Yeah, I think it would be better if they live somewhere in `utils`. I have put some template related stuff in `concepts_lite.hpp` in the past, but this might be enough to make a new header in utils called `type_lists.hpp` or something.

Reply:
I ended up adding some type list functionality in `utils/type_list.hpp` in #1143 for dealing with grid variable types. I am not sure how much overlap there is with what you are doing there, but at least that could be a place to put this stuff (at least once #1143 is in develop).

Reply:
Nice! There is some overlap but I think your use of tuples looks nicer than all the recursive templates I used